File: Python/graphs/Rotten_Oranges.py
'''
Problem - 
	To find the time taken for all oranges to get rotten.

Algorithm - 
    We follow by DFS search starting from those points that have rotten oranges and finding thoses fresh oranges in adjacent to them.
    We keep a queue and as we get a new rotten orange we push it into the queue and do this until all possible oranges are rotten.
    The time taken is then returned.
	
'''

from queue import Queue
import copy

class rotten_Oranges():
    def __init__(self, arr):
        self.arr = copy.deepcopy(arr)
        self.R = len(arr)
        self.C = len(arr[0])
        self.time = -1
    
    # Checks if the index in consideration is valid or not
    def isValid(self, x, y):
        return (x >= 0 and x < self.R and y >= 0 and y < self.C)
    
    # Checks if there are still some fresh oranges left
    def checkall(self):
    	for i in range(self.R):
    		for j in range(self.C):
    			if (self.arr[i][j] == 1):
    				return True;
    	return False;
    
    # A checker to take care of all corner case. In case the matrix is all zero then this checker takes care of such cases
    def checkzero(self):
    	for i in range(self.R):
    		for j in range(self.C):
    			if (self.arr[i][j] != 0):
    				return False;
    	return True;

    # The main function where the calculation for time taken takes place
    def timeTaken(self):
        q = Queue(0)
        xindex = [1, -1, 0, 0]
        yindex = [0, 0, 1, -1]
        
        for i in range(self.R):
            for j in range(self.C):
                if (self.arr[i][j] == 2):
                    q.put([i, j])
        
        while (not q.empty()):
            self.time += 1
            lsize = q.qsize()
            
            for i in range(lsize):
                curr_x, curr_y = q.get()
                
                for j in range(4):
                    x = curr_x + xindex[j]
                    y = curr_y + yindex[j]
                
                    if (not self.isValid(x, y)):
                        continue
            
                    if (self.arr[x][y] == 1):
                        self.arr[x][y] = 2
                        q.put([x, y])
    
        if (self.checkzero()):
            return 0
	
        return self.time if(not self.checkall()) else -1

arr = []
R, C = map(int, input("Enter number of rows and columns : ").split())
for i in range(R):
    print(("\nEnter for {}th row : ").format(i+1), end = "")
    arr.append(list(map(int, input().split())))

obj = rotten_Oranges(arr)
time = obj.timeTaken()
if (time == -1):
    print("\nAll the oranges cannot be rotten !!")
else:
    print("\nIt takes {} seconds.".format(time))

'''
Test run example - 
	Enter number of rows and columns : 3 5
	Enter for 1th row : 2 1 0 2 1
	Enter for 2th row : 1 0 1 2 1
    	Enter for 3th row : 1 0 0 2 1
    
    	It takes 2 seconds.

Time and Space Complexity - 
	Time - O(R * C)
	Space - O(R * C)
'''

File: C-Plus-Plus/cp/Rotten_Oranges.cpp
/*
Each cell is represented by either

0: Empty cell
1: Cells have fresh oranges
2: Cells have rotten oranges

Only a fresh orange adjacent to a rotten one will get rotten after 1 second.
*/

#include <bits/stdc++.h>
using namespace std;

// A struct data type for storing each and every cell index pairs
struct cell
{
	int x, y;
};

// Checks if the index in consideration is valid or not
bool isValid(int x, int y, int R, int C)
{
	return (x >= 0 && x < R && y >= 0 && y < C);
}

// Checks if there are still some fresh oranges left
bool checkall(vector <int> arr[], int R)
{
	int C = arr[0].size();
	for (int i = 0; i < R; i++)
		for (int j = 0; j < C; j++)
			if (arr[i][j] == 1)
				return true;
	return false;
}

// A checker to take care of all corner case. In case the matrix is all zero then this checker takes care of such cases
bool checkzero(vector <int> arr[], int R)
{
	int C = arr[0].size();
    for (int i = 0; i < R; i++)
        for (int j = 0; j < C; j++)
            if (arr[i][j] != 0)
                return false;
    return true;
}

int rotOranges(vector <int> arr[], int R)
{
	queue <cell> q;
	cell curr;
	int time = -1, lsize, x, y, i, j;
	int xindex[] = {1, -1, 0, 0};
	int yindex[] = {0, 0, 1, -1};
	int C = arr[0].size();
	
	for (i=0; i < R; i++)
		for (j=0; j < C; j++)
			if (arr[i][j] == 2)
				q.push({i, j});

	while (!q.empty())
	{
		time++;
		lsize = q.size();
		
		while (lsize--)
		{
			curr = q.front();
			q.pop();
			
			for (int i=0; i < 4; i++)
			{
				x = curr.x + xindex[i];
				y = curr.y + yindex[i];
				
				if (!isValid(x, y, R, C))
					continue;
				
				if (arr[x][y] == 1)
				{
					arr[x][y] = 2;
					q.push({x, y});
				}
			}
		}
	}
	
	if (checkzero(arr, R))
		return 0;
	
	if (checkall(arr, R))
		return -1;
	return time;
}

int main()
{
	int R, C, num;
	
	cout << "Enter number of rows and columns: ";
	cin >> R >> C;
	
	vector <int> arr[R];
	
	for (int i = 0; i < R; i++)
	{
		for (int j = 0; j < C; j++)
		{
			cout << "\nEnter number : ";
			cin >> num;
			arr[i].push_back(num);
		}
	}
	
	int ans = rotOranges(arr, 3);
	if (ans != -1)
		cout << "\nThe time taken is " << ans << " seconds.";
    else
		cout << "\nAll the oranges cannot be rotten !!";
	
    return 0;
}

/*
Row = 3
Column = 5

Time t = 0

	2, 1, 0, 2, 1
	1, 0, 1, 2, 1
	1, 0, 0, 2, 1

Time t = 1

	2, 2, 0, 2, 2
	2, 0, 2, 2, 2
	1, 0, 0, 2, 2

Time t = 2

	2, 2, 0, 2, 2
	2, 0, 2, 2, 2
	2, 0, 0, 2, 2

All oranges are rotten. Time taken is 2 seconds.
*/
